include/modem/at_parser.h

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#ifndef AT_PARSER_H__
+#define AT_PARSER_H__
+
+#include <stdbool.h>
+#include <zephyr/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+enum at_num_type {
+	AT_NUM_INT16,
+	AT_NUM_UINT16,
+	AT_NUM_INT32,
+	AT_NUM_UINT32,
+	AT_NUM_INT64,
+};
+
+enum at_cmd_type {
+	AT_CMD_UNKNOWN,
+	AT_CMD_SET_COMMAND,
+	AT_CMD_READ_COMMAND,
+	AT_CMD_TEST_COMMAND
+};
+
+struct at_parser {
+	const char *cursor;
+	size_t count;
+	bool is_next_empty;
+	bool is_init;
+};
+
+int at_parser_init(struct at_parser *parser, const char *at);
+int at_parser_num_get(struct at_parser *parser, size_t index, void *value, enum at_num_type type);
```

Comment:
here instead of using a `enum` to determine what to copy, I would check if we can use C11 generics or alternatively, let the parameter be a [transparent union](https://gcc.gnu.org/onlinedocs/gcc-8.3.0/gcc/Common-Type-Attributes.html) of all the supported types.

Reply:
Didn't think about that, neat idea.